File: engine/promptectomy/contracts.py
"""Frozen contracts for PROMPTECTOMY. Do not change without the integrator (main owner).

These Pydantic models are the single source of truth for:
  - the JSONL ledger of recorded LLM calls (LedgerEventV1),
  - the callsite audit that Codex emits under --output-schema (AuditReport),
  - the per-callsite Verdict the PARENT verifier emits (never Codex),
  - the PipelineEvent stream written to NDJSON and consumed by the dashboard.

The PipelineEvent JSON shape is mirrored exactly in ui/lib/contracts.ts (Zod). Same wire shape,
two validators. The engine emits PipelineEvents; the Next route validates and forwards them.
"""

from __future__ import annotations

from typing import Annotated, Literal, Union

from pydantic import BaseModel, ConfigDict, Field, JsonValue


class Closed(BaseModel):
    model_config = ConfigDict(extra="forbid")


# --- Ledger (recording of real LLM calls; internal to the engine) ---------------------

CallKind = Literal["structured", "classifier", "freeform"]
CostSource = Literal["llm", "compiled"]


class Usage(Closed):
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0


class LedgerEventV1(Closed):
    schema_version: Literal["1"] = "1"
    event_id: str
    recorded_at: str
    repo_sha: str
    callsite_id: str
    mode: Literal["original", "compiled", "shadow"] = "original"
    api: Literal["responses.create", "chat.completions.create"] = "responses.create"
    model: str
    # Never record Authorization, API keys, env, or raw headers.
    request_input: JsonValue
    request_params: dict[str, JsonValue] = Field(default_factory=dict)
    response_raw: JsonValue | None = None
    response_normalized: JsonValue | None = None
    usage: Usage = Field(default_factory=Usage)
    latency_ms: float = Field(ge=0)
    estimated_cost_usd: float | None = Field(default=None, ge=0)
    status: Literal["ok", "error"] = "ok"


# --- Callsite audit (Codex emits under --output-schema; verdict is NOT decided here) ---


class AuditSample(Closed):
    # str, not JsonValue: these are display previews, and OpenAI strict --output-schema rejects
    # a field with no concrete type (JsonValue), which silently forced the scanner onto its fallback.
    event_id: str
    input_preview: str = ""
    output_preview: str = ""


class CallsiteAudit(Closed):
    callsite_id: str
    file: str
    line: int = Field(ge=1)
    symbol: str = ""
    kind: CallKind  # structured | classifier | freeform (UI-facing simplification)
    eligibility: Literal["candidate", "keep_model", "unsupported"]
    sample_count: int = Field(ge=0)
    confidence: float = Field(ge=0, le=1, default=0.5)
    reason: str = ""
    samples: list[AuditSample] = Field(default_factory=list, max_length=3)


class AuditReport(Closed):
    schema_version: Literal["1"] = "1"
    repo_path: str
    repo_sha: str
    callsites: list[CallsiteAudit]


# --- Verdict (the PARENT verifier emits this after the sealed holdout; Codex never does) ---

VerdictStatus = Literal["COMPILED", "COMPILED_WITH_DIFFS", "NOT_COMPILABLE", "ERROR"]


class Diff(Closed):
    event_id: str = ""
    input: JsonValue
    expected: JsonValue
    got: JsonValue
    summary: str = ""


class Verdict(Closed):
    schema_version: Literal["1"] = "1"
    callsite_id: str
    status: VerdictStatus
    reason: str = ""
    agreement: float | None = Field(default=None, ge=0, le=1)
    holdout_n: int = Field(ge=0, default=0)
    diffs: list[Diff] = Field(default_factory=list)
    module_path: str | None = None
    median_latency_ms: float | None = None


# --- PipelineEvent stream (NDJSON; consumed by the dashboard). Mirror of contracts.ts ---


class ScanEvent(Closed):
    type: Literal["scan"] = "scan"
    audit: list[CallsiteAudit]
    recorded_calls: int


class TrafficEvent(Closed):
    type: Literal["traffic"] = "traffic"
    callsite_id: str
    latency_ms: float
    cost_usd: float
    source: CostSource


class CallsiteStatusEvent(Closed):
    type: Literal["callsiteStatus"] = "callsiteStatus"
    callsite_id: str
    status: Literal["queued", "synthesizing", "replaying", "iterating", "done"]
    round: int | None = None


class SynthTokenEvent(Closed):
    type: Literal["synthToken"] = "synthToken"
    callsite_id: str
    text: str
    kind: Literal["reasoning", "code", "tool"] | None = None


class ReplayEvent(Closed):
    type: Literal["replay"] = "replay"
    callsite_id: str
    passed: int
    total: int


class VerdictEvent(Closed):
    type: Literal["verdict"] = "verdict"
    verdict: Verdict


class SwapSample(Closed):
    input: JsonValue
    llm_output: JsonValue
    compiled_output: JsonValue


class SwapEvent(Closed):
    type: Literal["swap"] = "swap"
    callsite_id: str
    before_ms: float
    after_ms: float
    before_cost: float
    samples: list[SwapSample] = Field(default_factory=list)


class DoneTotals(Closed):
    callsites: int
    compiled: int
    kept: int
    est_monthly_savings_usd: float
    pipeline_cost_reduction_pct: float


class DoneEvent(Closed):
    type: Literal["done"] = "done"
    totals: DoneTotals


PipelineEvent = Annotated[
    Union[
        ScanEvent,
        TrafficEvent,
        CallsiteStatusEvent,
        SynthTokenEvent,
        ReplayEvent,
        VerdictEvent,
        SwapEvent,
        DoneEvent,
    ],
    Field(discriminator="type"),
]

File: ui/lib/contracts.ts
// Frozen contracts for the PROMPTECTOMY dashboard. Mirror of engine/promptectomy/contracts.py.
// Same NDJSON wire shape, two validators (Pydantic on the engine, Zod here). The engine emits
// PipelineEvents to NDJSON; the Next route validates each line against these schemas and forwards.
// Do not change without the integrator (main owner).

import { z } from "zod";

export const CallKind = z.enum(["structured", "classifier", "freeform"]);

export const CallsiteAudit = z.object({
  callsite_id: z.string(),
  file: z.string(),
  line: z.number().int().min(1),
  symbol: z.string().default(""),
  kind: CallKind,
  eligibility: z.enum(["candidate", "keep_model", "unsupported"]),
  sample_count: z.number().int().min(0),
});
export type CallsiteAudit = z.infer<typeof CallsiteAudit>;

export const Diff = z.object({
  event_id: z.string().default(""),
  input: z.unknown(),
  expected: z.unknown(),
  got: z.unknown(),
  summary: z.string().default(""),
});

export const Verdict = z.object({
  callsite_id: z.string(),
  status: z.enum(["COMPILED", "COMPILED_WITH_DIFFS", "NOT_COMPILABLE", "ERROR"]),
  reason: z.string().default(""),
  agreement: z.number().min(0).max(1).nullable().default(null),
  holdout_n: z.number().int().min(0).default(0),
  diffs: z.array(Diff).default([]),
  module_path: z.string().nullable().default(null),
  median_latency_ms: z.number().nullable().default(null),
});
export type Verdict = z.infer<typeof Verdict>;

const Scan = z.object({
  type: z.literal("scan"),
  audit: z.array(CallsiteAudit),
  recorded_calls: z.number().int(),
});
const Traffic = z.object({
  type: z.literal("traffic"),
  callsite_id: z.string(),
  latency_ms: z.number(),
  cost_usd: z.number(),
  source: z.enum(["llm", "compiled"]),
});
const CallsiteStatus = z.object({
  type: z.literal("callsiteStatus"),
  callsite_id: z.string(),
  status: z.enum(["queued", "synthesizing", "replaying", "iterating", "done"]),
  round: z.number().int().nullable().optional(),
});
const SynthToken = z.object({
  type: z.literal("synthToken"),
  callsite_id: z.string(),
  text: z.string(),
  kind: z.enum(["reasoning", "code", "tool"]).nullable().optional(),
});
const Replay = z.object({
  type: z.literal("replay"),
  callsite_id: z.string(),
  passed: z.number().int(),
  total: z.number().int(),
});
const VerdictEvent = z.object({ type: z.literal("verdict"), verdict: Verdict });
const SwapSample = z.object({
  input: z.unknown(),
  llm_output: z.unknown(),
  compiled_output: z.unknown(),
});
const Swap = z.object({
  type: z.literal("swap"),
  callsite_id: z.string(),
  before_ms: z.number(),
  after_ms: z.number(),
  before_cost: z.number(),
  samples: z.array(SwapSample).default([]),
});
const Done = z.object({
  type: z.literal("done"),
  totals: z.object({
    callsites: z.number().int(),
    compiled: z.number().int(),
    kept: z.number().int(),
    est_monthly_savings_usd: z.number(),
    pipeline_cost_reduction_pct: z.number(),
  }),
});

export const PipelineEvent = z.discriminatedUnion("type", [
  Scan,
  Traffic,
  CallsiteStatus,
  SynthToken,
  Replay,
  VerdictEvent,
  Swap,
  Done,
]);
export type PipelineEvent = z.infer<typeof PipelineEvent>;

// Recorder/replayer wrapper (NOT the wire shape): one NDJSON line per recorded event.
export const RecordedLine = z.object({ at_ms: z.number(), event: PipelineEvent });
export type RecordedLine = z.infer<typeof RecordedLine>;
